decider-bcl2fastq/src/main/java/ca/on/oicr/pde/deciders/utils/BarcodeAndBasesMask.java

```diff
@@ -0,0 +1,178 @@
+            barcodeOne = null;
+        } else if (basesMask.getIndexOneIncludeLength() == null && barcodeOne != null) {
+            barcodeOne = null;
+        } else if (basesMask.getIndexOneIncludeLength() != null && barcodeOne == null) {
+            if (strict) {
+                throw new DataMismatchException("Barcode one missing but bases mask requires it");
+            }
+        } else if (basesMask.getIndexOneIncludeLength() != null && barcodeOne != null) {
+            if (barcodeOne.length() < basesMask.getIndexOneIncludeLength()) {
+                // do not modify barcode
+            } else {
+                barcodeOne = barcodeOne.substring(0, basesMask.getIndexOneIncludeLength());
+            }
+        }
+
+        if (basesMask.getIndexTwoIncludeLength() == null && barcodeTwo == null) {
+            barcodeTwo = null;
+        } else if (basesMask.getIndexTwoIncludeLength() == null && barcodeTwo != null) {
+            barcodeTwo = null;
+        } else if (basesMask.getIndexTwoIncludeLength() != null && barcodeTwo == null) {
+            if (strict) {
+                throw new DataMismatchException("Barcode two missing but bases mask requires it");
+            }
+        } else if (basesMask.getIndexTwoIncludeLength() != null && barcodeTwo != null) {
+            if (barcodeTwo.length() < basesMask.getIndexTwoIncludeLength()) {
+                // do not modify barcode
+            } else {
+                barcodeTwo = barcodeTwo.substring(0, basesMask.getIndexTwoIncludeLength());
+            }
+        }
+
+        if (barcodeOne != null && !barcodeOne.isEmpty() && (barcodeTwo == null || barcodeTwo.isEmpty())) {
+            return new Barcode(barcodeOne);
+        } else if ((barcodeOne == null || barcodeOne.isEmpty()) && barcodeTwo != null && !barcodeTwo.isEmpty()) {
+            return new Barcode(barcodeTwo);
+        } else if (barcodeOne != null && !barcodeOne.isEmpty() && barcodeTwo != null && !barcodeTwo.isEmpty()) {
+            return new Barcode(barcodeOne, barcodeTwo);
+        } else {
+            throw new DataMismatchException("Unexpected state, barcode one = [" + barcodeOne + "], barcode two = [" + barcodeTwo + "]");
+        }
+
+    }
+
+    public static BasesMask calculateBasesMask(Barcode barcode) {
+
+        BasesMask.BasesMaskBuilder basesMaskBuilder = new BasesMask.BasesMaskBuilder();
+
+        if (barcode.getBarcodeOne().length() > 0) {
+            basesMaskBuilder.setIndexOneIncludeLength(barcode.getBarcodeOne().length());
+        }
+
+        if (barcode.getBarcodeTwo() != null && barcode.getBarcodeTwo().length() > 0) {
+            basesMaskBuilder.setIndexTwoIncludeLength(barcode.getBarcodeTwo().length());
+            basesMaskBuilder.setIndexTwoIgnoreLength(Integer.MAX_VALUE);
+        }
+
+        return basesMaskBuilder.createBasesMask();
+    }
+
+    public static BasesMask calculateBasesMask(Barcode barcode, BasesMask runBasesMask) throws DataMismatchException {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 21 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)